File: SWEA/D4/5643. ［Professional］ 키 순서/［Professional］ 키 순서.java
import java.io.*;
import java.util.*;

public class Solution {
  static int N, M;
  static List<List<Integer>> sgraph;
  static List<List<Integer>> bgraph;

  static int bfs(int s, boolean isSmall) {
    boolean [] visited = new boolean[N + 1];
    Queue<Integer> nv = new ArrayDeque<>();

    int cnt = 0;
    visited[s] = true;
    nv.add(s);

    while (!nv.isEmpty()) {
      int cur = nv.poll();

      if (isSmall) { // s보다 작은 사람 탐색
        for (int nxt : bgraph.get(cur)) { // bgraph를 이용해 cur보다 작은 사람 탐색
          if (!visited[nxt]) {
            visited[nxt] = true;
            nv.add(nxt);
            cnt++;
          }
        }
      } else { // s보다 큰 사람 탐색
        for (int nxt : sgraph.get(cur)) { // sgraph를 이용해 cur보다 큰 사람 탐색
          if (!visited[nxt]) {
            visited[nxt] = true;
            nv.add(nxt);
            cnt++;
          }
        }
      }
    }
    return cnt;
  }
  public static void main (String args[]) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringBuilder sb = new StringBuilder();
    int T = Integer.parseInt(br.readLine());

    for (int tc = 1; tc <= T; tc++) {
      // 입력
      N = Integer.parseInt(br.readLine()); // 학생들의 수
      M = Integer.parseInt(br.readLine()); // 학생의 키를 비교한 횟수

      // 초기화
      sgraph = new ArrayList<>();
      bgraph = new ArrayList<>();

      for (int i = 0; i <= N; i++) {
        sgraph.add(new ArrayList<>());
        bgraph.add(new ArrayList<>());
      }

      for (int i = 0; i < M; i++) {
        // 번호가 a인 학생이 번호가 b인 학생보다 키가 작다 (유향그래프)
        StringTokenizer st = new StringTokenizer(br.readLine());
        int a = Integer.parseInt(st.nextToken());
        int b = Integer.parseInt(st.nextToken());
        sgraph.get(a).add(b); // a보다 큰 학생들 리스트
        bgraph.get(b).add(a); // b보다 작은 학생들 리스트
      }

      int ans = 0;
      for (int i = 1; i <= N; i++) {
        int smaller = bfs(i, true); // i보다 작은 사람의 수
        int bigger = bfs(i, false); // i보다 큰 사람의 수

        // 둘의 합이 전체 명수에서 자기 자신을 뺀 값이어야 모든 사람과 키 비교 가능
        if (smaller + bigger == N - 1) ans++;
      }
      sb.append("#").append(tc).append(" ").append(ans).append("\n");
    }
    System.out.println(sb);
  }
}

File: SWEA/D4/5643. ［Professional］ 키 순서/［Professional］ 키 순서.py
from collections import deque

def bfs(s, is_small):
  visited = [False] * (n+1)
  nv = deque()

  cnt = 0
  visited[s] = True
  nv.append(s)

  while nv:
    cur = nv.pop()

    if is_small: # s보다 작은 사람 탐색
      for nxt in bgraph[cur]: # bgraph는 cur보다 작은 사람들
        if not visited[nxt]:
          visited[nxt] = True
          nv.append(nxt)
          cnt += 1
    else: # s보다 큰 사람 탐색
      for nxt in sgraph[cur]: # sgraph는 cur보다 큰 사람들
        if not visited[nxt]:
          visited[nxt] = True
          nv.append(nxt)
          cnt += 1
  
  return cnt

for tc in range (1, int(input()) + 1):
  n = int(input())
  m = int(input())
  sgraph = {i:[] for i in range (1, n + 1)} # i보다 []에 있는 사람이 큼
  bgraph = {i:[] for i in range (1, n + 1)} # i보다 []에 있는 사람이 작음

  for _ in range (m):
    # a의 키보다 b의 키가 큼
    a, b = map(int, input().split())
    sgraph[a].append(b) # a보다 큰 사람들
    bgraph[b].append(a) # b보다 작은 사람들
  
  ans = 0
  for i in range (1, n + 1):
    smaller = bfs(i, True) # i보다 작은 사람의 수
    bigger = bfs(i, False) # i보다 큰 사람의 수

    # i보다 작은 사람과 i보다 큰 사람을 더했을 때 n-1이면 모든 사람과 키 비교가 가능
    if (smaller + bigger == n - 1):
      ans += 1
    
  print(f"#{tc} {ans}")